libopae/plugins/xfpga/metrics/afu_metrics.c

```diff
@@ -0,0 +1,275 @@
+// Copyright(c) 2018, Intel Corporation
+//
+// Redistribution  and  use  in source  and  binary  forms,  with  or  without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of  source code  must retain the  above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/**
+* \file afu_metrics.c
+* \brief Enumerates AFU Metrics BBB & retrives afu metrics values
+*/
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif // HAVE_CONFIG_H
+
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <dirent.h>
+#include <uuid/uuid.h>
+
+#include "xfpga.h"
+#include "common_int.h"
+#include "metrics_int.h"
+#include "types_int.h"
+#include "opae/metrics.h"
+#include "metrics/vector.h"
+#include "safe_string/safe_string.h"
+
+// AFU BBB GUID
+#define METRICS_BBB_GUID            "87816958-C148-4CD0-9D73-E8F258E9E3D7"
+#define METRICS_BBB_ID_H            0x87816958C1484CD0
+#define METRICS_BBB_ID_L            0x9D73E8F258E9E3D7
+
+#define FEATURE_TYPE_BBB            0x2
+
+#define METRIC_CSR_OFFSET           0x20
+#define METRIC_NEXT_CSR             0x8
+
+
+fpga_result discover_afu_metrics_feature(fpga_handle handle, uint64_t *offset)
+{
+	fpga_result result               = FPGA_OK;
+	feature_definition feature_def;
+	uint64_t bbs_offset              = 0;
+
+	memset_s(&feature_def, 0, sizeof(feature_def));
```

Comment:
The signature of memset_s is..
```c
errno_t memset_s(void *dest, rsize_t len, uint8_t value);
```
This call isn't correct.  Please fix.

Reply:
fixed